controllers/machinehealthcheck_controller_test.go

```diff
@@ -111,841 +110,968 @@ func newInfraMachine(machine *clusterv1.Machine) (*unstructured.Unstructured, st
 	}, providerID
 }
 
-var _ = Describe("MachineHealthCheck", func() {
-	Context("on reconciliation", func() {
-		var (
-			cluster *clusterv1.Cluster
-			mhc     *clusterv1.MachineHealthCheck
-		)
-
-		var (
-			nodes             []*corev1.Node
-			machines          []*clusterv1.Machine
-			fakeNodesMachines = func(n int, healthy bool, nodeRef bool) {
-
-				for i := 0; i < n; i++ {
-					machine := newMachine(mhc, cluster)
-					infraMachine, providerID := newInfraMachine(&machine)
-					node := newNode()
-					if !healthy {
-						setNodeUnhealthy(&node)
-					}
-					Expect(testEnv.Create(ctx, infraMachine)).To(Succeed())
-					fmt.Fprintf(GinkgoWriter, "inframachine created: %s\n", infraMachine.GetName())
-					machine.Spec.InfrastructureRef = corev1.ObjectReference{
-						APIVersion: infraMachine.GetAPIVersion(),
-						Kind:       infraMachine.GetKind(),
-						Name:       infraMachine.GetName(),
-					}
-					infraMachinePatch := client.MergeFrom(infraMachine.DeepCopy())
-					Expect(unstructured.SetNestedField(infraMachine.Object, true, "status", "ready")).To(Succeed())
-					Expect(testEnv.Status().Patch(ctx, infraMachine, infraMachinePatch)).To(Succeed())
-
-					machineStatus := machine.Status
-					Expect(testEnv.Create(ctx, &machine)).To(Succeed())
-					fmt.Fprintf(GinkgoWriter, "machine created: %s\n", machine.GetName())
-
-					if nodeRef {
-						node.Spec.ProviderID = providerID
-						nodeStatus := node.Status
-						Expect(testEnv.Create(ctx, &node)).To(Succeed())
-						fmt.Fprintf(GinkgoWriter, "node created: %s\n", node.GetName())
-						nodePatch := client.MergeFrom(&node)
-						node.Status = nodeStatus
-						Expect(testEnv.Status().Patch(ctx, &node, nodePatch)).To(Succeed())
-						nodes = append(nodes, &node)
-
-						machinePatch := client.MergeFrom(machine.DeepCopy())
-						machine.Status = machineStatus
-						machine.Status.NodeRef = &corev1.ObjectReference{
-							Name: node.Name,
-						}
-
-						now := metav1.NewTime(time.Now())
-						machine.Status.LastUpdated = &now
-						Expect(testEnv.Status().Patch(ctx, &machine, machinePatch)).To(Succeed())
-					}
+func TestMachineHealthCheck_Reconcile(t *testing.T) {
```

Comment:
This is a big diff. But essentially, I'm pulling out each `Specify` block into a separate `t.Run` subtest and isolating all the objects it requires so as to minimize shared state.